experimental/Schemes/singular_locus.jl

```diff
@@ -0,0 +1,147 @@
+    end
+  end
+  return result
+end
+
+@attr Bool function is_equidimensional(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  P = primary_decomposition(saturated_ideal(I))
+  length(P) == 0 && return true
+  d = dim(P[1][1])
+  all(x->dim(x[1])==d, P[2:end]) && return true
+  return false
+end
+
+@attr function reduced_scheme(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  J = radical(saturated_ideal(I))
+  return Spec(base_ring(J), J, inverted_set(OO(X)))
+end
+
+@attr function is_reduced(X::AbsSpec{<:Ring, <:MPolyQuoLocalizedRing})
+  I = localized_modulus(OO(X))
+  J = saturated_ideal(I)
+  return is_reduced(quo(base_ring(OO(X)), J)[1])
+end
+
+
+########################################################################
+# Additional functionality for fractions                               #
+########################################################################
+
+function derivative(f::MPolyLocalizedRingElem, i::Int)
+  num = derivative(numerator(f), i)*denominator(f) - derivative(denominator(f), i)*numerator(f)
+  den = denominator(f)^2
+  g = gcd(num, den)
+  return parent(f)(divexact(num, g), divexact(den, g), check=false)
+end
+
+function derivative(f::MPolyQuoLocalizedRingElem, i::Int)
+  num = derivative(lifted_numerator(f), i)*lifted_denominator(f) - derivative(lifted_denominator(f), i)*lifted_numerator(f)
+  den = lifted_denominator(f)^2
+  g = gcd(num, den)
+  return parent(f)(divexact(num, g), divexact(den, g), check=false)
+end
+
+function jacobi_matrix(f::MPolyLocalizedRingElem)
+  L = parent(f)
+  n = nvars(base_ring(L))
+  return matrix(L, n, 1, [derivative(f, i) for i=1:n])
+end
+
+function jacobi_matrix(g::Vector{<:MPolyLocalizedRingElem})
+  R = parent(g[1])
+  n = nvars(base_ring(R))
+  @assert all(x->parent(x) == R, g)
+  return matrix(R, n, length(g), [derivative(x, i) for i=1:n for x = g])
+end
+
+function jacobi_matrix(f::MPolyQuoLocalizedRingElem)
+  L = parent(f)
+  n = nvars(base_ring(L))
+  return matrix(L, n, 1, [derivative(f, i) for i=1:n])
+end
+
+function jacobi_matrix(g::Vector{<:MPolyQuoLocalizedRingElem})
+  L = parent(g[1])
+  n = nvars(base_ring(L))
+  @assert all(x->parent(x) == L, g)
+  return matrix(L, n, length(g), [derivative(x, i) for i=1:n for x = g])
+end
+
+########################################################################
+# Smoothness test based on projective modules.
+#
+# The routine checks whether the module for the cotangent sheaf Ω¹(X)
+# is locally free over 𝒪(X) and returns `true` if this is the case. 
+########################################################################
+
+@attr Bool function is_smooth(X::AbsSpec{<:Field, <:MPolyQuoLocalizedRing})
+  R = base_ring(OO(X))
+  characteristic(base_ring(X)) == 0 || error("method not implemented in positive characteristic")
+  L = localized_ring(OO(X))
+  I = localized_modulus(OO(X))
+  f = gens(I)
+  Df = jacobi_matrix(f)
+
+  A = map_entries(x->OO(X)(x), Df)
```

Comment:
Remark referring to a previous comment: Well, this is precisely such an application of the jacobian matrix, where your subsequent computations are modulo the ideal anyway and you can simply forget about the denominators without messing up the result.